Strip blank tags out from metadata lists

afontenot opened the pull request:
When a tag is an empty string, return an empty list instead.

It's common to have a tag in a file be present but blank. When
this happens, lists of metadata used in collections (e.g. by the
~persons tag) will contain empty strings and be combined and exposed
to the user in ugly ways. (Example: "Artist 1, , Artist 2,")

Fixing this doesn't appear to have an negative side effects.

Fixes #3383

afontenot commented on quodlibet/formats/_audio.py:
```diff
@@ -743,7 +743,7 @@ def list(self, key):
         else:
             v = self.get(key)
-            return [] if v is None else v.split("\n")
+            return [] if v in (None, "") else v.split("\n")
```
I didn't want to assume that the metadata dict couldn't have `0` as a value for some piece of metadata. If the devs say this is supposed to always be a string, I'll change it.

jej replied:
Should be ok according to line 112:
```
    Keys are either ASCII str or unicode.
    Values are always unicode except if the tag is part of FILESYSTEM_TAGS,
    then the value is of the path type (str on UNIX, unicode on Windows)
```

https://github.com/quodlibet/quodlibet/blob/9509cd2a8087e72884d847e4a63046bad513dd20/quodlibet/formats/_audio.py#L112

declension commented:
It seems good generally, but I worry it might hide the ability for users to _find_ these empty tags (and, say, remove them) - have you checked this @afontenot ?

Also, please add unit tests for any new behaviour.

afontenot commented:
I've added a test and made the change suggested by @jej. I've force-pushed the changes to keep everything cleanly in one commit and as of this writing the Github interface hasn't caught up.

> It seems good generally, but I worry it might hide the ability for users to _find_ these empty tags (and, say, remove them) - have you checked this @afontenot ?

Yes, I've done some basic checks. It's not immediately obvious to me what all the callers of `.list()` are, so there could be corner cases I haven't considered, but it seems like it's not used (and probably not supposed to be used) to determine whether or not a tag *exists* - which means that making this change will never hide the presence of the tag. I have verified, for example, that the built in tag editor still shows the blank tag. This fixes several places in which the blank tag is unexpectedly exposed to the user, e.g. when combined in the people tag or in the information window.

declension commented on tests/test_formats__audio.py:
```diff
@@ -737,6 +737,12 @@ def test_peoplesort(self):
         self.failUnlessEqual(q.list("~peoplesort:roles"),
             ["B, The (Guitar)", "C, The (Performance)", "A, The (Vocals)"])
 
+    def test_blanktaghandling(self):
```
```suggestion
    def test_blank_tag_handling(self):
```

declension commented:
> I have verified, for example, that the built in tag editor still shows the blank tag.

Thanks! This was my main concern, so I think we're good to go now

afontenot commented on tests/test_formats__audio.py:
```diff
@@ -737,6 +737,12 @@ def test_peoplesort(self):
 
+    def test_blanktaghandling(self):
+        q = AudioFile([("artist", "A"),
+                        ("performer", "")])
```
Okay, I've had time to look into this for real, and it is a potential problem. I was able to add multiple "artist" tags to a file, one of them blank, and it exemplifies that problem in the UI as well. So I'll change the PR to cover this case as well and add it to the test case as well. If you meant something else, let me know and I'll try to work on that.

afontenot commented:
So, there's at least one additional issue that I just discovered. Sometimes consumers of AudioFile will just call `comma()` and use that instead of operating on the list. And `comma` just replaces the `"\n"` in the tag with `", "`. So this should probably be fixed there as well.

There's also probably instances in which `list_sort()` would be affected. Possibly stupid question: `list_sort()` claims to return `display, sort` pairs. But isn't that just assuming that the order in which the tags are stored in the file is meaningful? i.e. 

```python3
AudioFile([("artist", "A\nB"),
           ("artistsort", "A\nB")])
```

I'm not sure what I can do about `list_sort()` unless I understand the assumption being made here.

Also, another stupid question. Why is AudioFile using `"\n"` delimited strings for tags to begin with? You're starting out with a list, e.g. in `MutagenVCFile`:

```python3
for key, value in (audio.tags or {}).items():
    self[key] = "\n".join(value)
```

The result of this is that every piece of code that pulls a tag has to do something with the newline characters, and many of them seem to make different assumptions for no obvious reason. Would a PR for storing tags with lists internally be worth considering?

declension commented:
> Okay, I've made several changes.
> 
>     * The name of the test, as you suggested (I was cribbing off the one right above mine, which isn't in snake_case).
Thanks, yes ideally they'd all get updated but it's too big / unimportant a task, so generally we try to tidy up new ones

>
>     * I've changed the behavior to always strip out blank tags from the list, even in cases where there are multiple instances of a tag and only one of them is blank.
Again so long as people can _fix_ their dodgy blank tags I'm OK

>     * I updated the docstring of the list function to be clear about the new behavior.
:+1: 

>     * I updated the test to cover behavior when there are multiple tags with the same name and make sure that ordering stays intact.
Great thanks


> I went with `list(filter(None` instead of a list comprehension because it seemed possibly more readable here and I saw that style used elsewhere in the code.

Yes - although again trying to replace those gradually (and sometimes less) for the more modern (relatively!) comprehension style generally, but no worries.

afontenot commented:
> Again so long as people can *fix* their dodgy blank tags I'm OK

Yes, I've checked that the blank tags still appear in the tag editor as before. So I think this could be merged as is, the only remaining questions are:

1. Do we want to also remove blank tags from the "comma" function? What about `list_sort()`?
2. Would it be worthwhile for me to look at replacing the current tag internal format (newline delimited strings) with lists? It's true there are some cases that only one instance of a tag makes sense (title?), but even in that case we're making an assumption and it might be good to have that assumption explicit. (I would make this a separate PR, not cram it into this one.)

declension commented:
> * Do we want to also remove blank tags from the "comma" function? What about `list_sort()`?
Probably yes :+1: 
In fact, it might be worth delegating `comma` to `list` (I tried this recently) except for the performance hit, so probably just repeating the logic slightly is better here.  

> * Would it be worthwhile for me to look at replacing the current tag internal format (newline delimited strings) with lists? It's true there are some cases that only one instance of a tag makes sense (title?), but even in that case we're making an assumption and it might be good to have that assumption explicit. (I would make this a separate PR, not cram it into this one.)

Maybe, but not before the next release anyway (whenever that is)

afontenot commented:
> > * Do we want to also remove blank tags from the "comma" function? What about `list_sort()`?
> 
> Probably yes +1 In fact, it might be worth delegating `comma` to `list` (I tried this recently) except for the performance hit, so probably just repeating the logic slightly is better here.

This might depend on what "repeating the logic" entails. I agree that a list would be cleaner - part of the reason I was asking about using lists internally is that it would avoid round-trips like this. Currently, I've gone with a regex, because it needs to remove any number of `\n` characters. Other than a for loop over the string (or accepting the round trip to list), I don't see any other reasonable implementation. I can switch it to round-trip through a list if you want.

I've also dealt with the `list_sort` function and added tests for both `list_sort` and `comma`. You should have a look at the second test for `list_sort` before merging this. Fixing the issue for `list_sort` required a slight change in behavior: before, for an AudioFile like

```python
s8 = {'artist': u'Artist1\n\nArtist3',
      'artistsort': u'SortA1\nSortA2'}
```

The result for `list_sort` is

```python
[('Artist1', 'SortA1'), ('', 'SortA2'), ('Artist3', 'Artist3')]
```

Now, we remove the middle artist entirely, and `'SortA2'` doesn't get used at all. Given that we're deliberately stripping out blank artists from the lists returned by this function, I think that's the only reasonable thing to do. Surely it wouldn't make sense to match up `'SortA2'` with `'Artist3'`! I've fixed several tests that failed after making this change.

I want to heavily **flag** the fact that I don't fully understand what `Pattern.format_list` is intended to do. I'm working on the assumption that it doesn't make sense to have blank tags floating around *anywhere*, but I don't 100% know that to be the case. I **have** checked that the tag editor continues to show blank tags after these changes.

declension commented:
I think this looks good to go. We can always fix up any weirdnesses users report.

declension commented:
Thanks @afontenot
